File: gui/src/features/ecommerce/api/checkoutApi.ts
import { httpClient } from '@shared/api/httpClient';
import { buildQueryString } from '@shared/utils/queryString';
import { CheckoutAddressInput, CheckoutPreview, OrderConfirmation } from '../types';

type ShowError = (msg: string) => void;

/**
 * Fronts `ecommerce-service`'s `CheckoutApi` (`/api/v1/checkout`, US-2.5–2.7), authenticated-only
 * through `gateway`. Two-step flow, same as the backend: `preview` revalidates the current cart
 * and shows what confirming right now would produce; `confirm` creates the order. Neither call
 * needs the caller to pass cart contents — both operate on the caller's own server-side cart.
 *
 * `selectedVariantIds` (post-Epic-2 follow-up) restricts either call to a subset of the cart;
 * omitted operates on the whole cart, same as before this parameter existed.
 *
 * `subtotalCouponCode`/`shippingCouponCode` (Coupon feature Phase 4) mirror `CheckoutApi`'s own
 * optional params — one code per `CouponTarget`, `undefined` applies none. `preview` never
 * redeems (safe to call repeatedly while a shopper tries different codes); `confirm` is what
 * actually consumes a redemption slot, only for whichever codes are passed at that moment.
 */
export const checkoutApi = {
  preview(
    selectedVariantIds?: number[],
    subtotalCouponCode?: string,
    shippingCouponCode?: string,
    showError?: ShowError,
  ): Promise<CheckoutPreview> {
    // Repeated query param (?selectedVariantIds=1&selectedVariantIds=2) — preview is a GET, so
    // there's no body to carry a selection in; matches CheckoutApi.preview's own @RequestParam.
    return httpClient.get(
      `/api/v1/checkout/preview${buildQueryString({ selectedVariantIds, subtotalCouponCode, shippingCouponCode })}`,
      showError,
    );
  },

  /** AddressBook follow-up: `address` carries either an existing saved address
   * (`savedAddressId`, every other field ignored) or a fresh, one-off address — see
   * `CheckoutAddressInput`'s own doc comment. */
  confirm(
    address: CheckoutAddressInput,
    selectedVariantIds?: number[],
    subtotalCouponCode?: string,
    shippingCouponCode?: string,
    showError?: ShowError,
  ): Promise<OrderConfirmation> {
    return httpClient.post(
      '/api/v1/checkout/confirm',
      { ...address, selectedVariantIds, subtotalCouponCode, shippingCouponCode },
      showError,
    );
  },
};

File: ecommerce-service/src/main/java/com/ttg/devknowledgeplatform/ecommerce/api/CheckoutApi.java
package com.ttg.devknowledgeplatform.ecommerce.api;

import com.ttg.devknowledgeplatform.common.annotation.CurrentUserId;
import com.ttg.devknowledgeplatform.ecommerce.dto.AddressRequest;
import com.ttg.devknowledgeplatform.ecommerce.dto.CheckoutConfirmResponse;
import com.ttg.devknowledgeplatform.ecommerce.dto.CheckoutPreviewResponse;

import jakarta.validation.Valid;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;

import java.util.List;

/**
 * HTTP contract for checkout (Epic 2, US-2.5–2.7). Authenticated-only, same as {@code CartApi} —
 * falls under this module's {@code security/SecurityConfig}'s default
 * {@code anyRequest().authenticated()} rule, no new rule needed.
 *
 * <p>Two-step flow: {@link #preview} for the shopper to review before committing to anything,
 * {@link #confirm} to actually place the order. See {@code CheckoutService}'s own Javadoc for why
 * {@code confirm} re-validates fresh rather than trusting a client-cached preview.
 */
@RequestMapping("/api/v1/checkout")
public interface CheckoutApi {

    /**
     * Revalidates the caller's cart and returns what confirming it right now would produce
     * (US-2.6).
     *
     * @param userUuid           the caller's Keycloak UUID
     * @param selectedVariantIds optional subset of variant ids to restrict this preview to
     *                           (repeated query param, e.g. {@code ?selectedVariantIds=1&selectedVariantIds=2});
     *                           omitted previews the whole cart
     * @param subtotalCouponCode optional coupon code targeting the subtotal (Coupon feature,
     *                           Phase 2); omitted/blank applies none
     * @param shippingCouponCode optional coupon code targeting the shipping fee; omitted/blank
     *                           applies none
     * @return {@code 200} with the reviewable preview
     */
    @GetMapping("/preview")
    ResponseEntity<CheckoutPreviewResponse> preview(
            @CurrentUserId String userUuid,
            @RequestParam(required = false) List<Integer> selectedVariantIds,
            @RequestParam(required = false) String subtotalCouponCode,
            @RequestParam(required = false) String shippingCouponCode);

    /**
     * Creates an order from the caller's currently-available cart lines and the given shipping
     * address (US-2.5, US-2.6), clearing the cart on success. {@code request} carries either an
     * existing AddressBook entry ({@code savedAddressId}) or a fresh, one-off address — see
     * {@code AddressRequest}'s own Javadoc for the full two-shape contract, including the optional
     * "save this for next time" flag.
     *
     * @param userUuid the caller's Keycloak UUID
     * @param request  the shipping address to snapshot onto the order
     * @return {@code 201} with the created order
     */
    @PostMapping("/confirm")
    ResponseEntity<CheckoutConfirmResponse> confirm(@CurrentUserId String userUuid, @Valid @RequestBody AddressRequest request);
}
